features/256-auth-gear-interaction-overview-in-next/api.md

```diff
@@ -0,0 +1,382 @@
+
+/* 
+ * user: user object to be saved
+ * orgUser: original user object
+ * currentUser: current exection user
+ */
+function before_XXX_sync(user, orgUser, currentUser) {
+    console.log(user.profile.loveCat); // false
+    
+    // alter user profile
+    user.profile.loveCat = true;
+    
+    /*
+     * or rasie exception
+     * throw new Error("some error");
+     */
+
+    // return updated user object
+    return user;
+}
+
+module.exports = skygear.auth.before_XXX_sync(before_XXX_sync);
+```
+
+Function signature of `after_XXX` hooked Cloud Function is:
+
+```javascript
+const skygear = require('skygear');
+
+/* 
+ * user: user object to be saved
+ * currentUser: current exection user
+ */
+function after_XXX(user, currentUser) {    
+    console.log(body); // { "loveCat": false }
+    console.log(user.profile.loveCat); // true
+    
+    /*
+     * or rasie exception
+     * throw new Error("some error");
+     */
+}
+
+module.exports = skygear.auth.after_XXX(after_XXX);
+```
+
+Followings are hooks of auth actions:
+
+| Action | Hooked Cloud Function |
+| -------- | -------- |
+| `signup` | `before_signup_sync(user, orgUser, currentUser)`<br/>`before_signup(user, orgUser, currentUser)`<br/>`after_signup_sync(user, currentUser)`<br/>`after_signup(user, currentUser)`<br/> |
+| `login` | `before_login_sync(user, orgUser, currentUser)`<br/>`before_login(user, orgUser, currentUser)`<br/>`after_login_sync(user, currentUser)`<br/>`after_login(user, currentUser)` |
+| `disable` | `before_disable_sync(user, orgUser, currentUser)`<br/>`before_disable(user, orgUser, currentUser)`<br/>`after_disable_sync(user, currentUser)`<br/>`after_disable(user, currentUser)` |
+| `role` | `before_change_role_sync(user, orgUser, currentUser)`<br/>`before_change_role(user, orgUser, currentUser)`<br/>`after_change_role_sync(user, currentUser)`<br/>`after_change_role(user, currentUser)` |
+| `logout` | `before_logout_sync(user, orgUser, currentUser)`<br/>`before_logout(user, orgUser, currentUser)`<br/>`after_logout_sync(user, currentUser)`<br/>`after_logout(user, currentUser)` |
+| `password` | `before_change_password_sync(user, orgUser, currentUser)`<br/>`before_change_password(user, orgUser, currentUser)`<br/>`after_change_password_sync(user, currentUser)`<br/>`after_change_password(user, currentUser)` |
+| `pasword` | `before_reset_password_sync(user, orgUser, currentUser)`<br/>`before_reset_password(user, orgUser, currentUser)`<br/>`after_reset_password_sync(user, currentUser)`<br/>`after_reset_password(user, currentUser)` |
+| `verify` | `before_verified_sync(user, orgUser, currentUser)`<br/>`before_verified(user, orgUser, currentUser)`<br/>`after_verified_sync(user, currentUser)`<br/>`after_verified(user, currentUser)` |
+| `update_user` | `before_update_user_sync(user, orgUser, currentUser)`<br/>`before_update_user(user, orgUser, currentUser)`<br/>`after_update_user_sync(user, currentUser)`<br/>`after_update_user(user, currentUser)` |
+
```

Comment:
@Steven-Chan 

1. In my opinion, "hook" reflects API invoking action rather than data field changes, so I don't think we need "un-verifying" hook (we don't have such API to "un-verifying" user). 
2. Same as the previous one, I think `reset_password` and `change_password` may act similarly,  they are different when they are invoked. Admin may want to trace them separately.

And follow the thought, I think `update_user` is only invoked when user metaData and profile updated, which implies auth gear needs to provide a new API `update_user` to support update user metaData and profile.

If we do want a trigger reflects data field changes (like v1 skygear record hook?), maybe we should define hook like: `change_verified`, `change_password`, `change_disabled`, ..., etc.  Want to know more about your opinions. Thanks.

Reply:
Take `last_login_at` as an example, a developer should consider creating `after_login` hook to handle requirements.

The reason for keeping `update_user_metadata_and_profile`(temp name) not triggering when auth data changed is that I want to avoid multiple hooks trigger for different reasons.

Take login as an example (`last_login_at` changed), following hooks would trigger:
`before_login_sync`
`before_update_user_sync`
`before_login`
`before_update_user`
`after_login_sync`
`after_update_user_sync`
`after_login`
`after_update_user`

A developer could feel confused where to implement application logic.

Reply:
Consider the major use case, if `last_login_at` or `created-at` is in profile DB of user, the user need to distribute code related with his profile DB in many hooks.

That seems even more confusing than if `update_user` combine all user object attribute update.

I don't really think having `update_user` hook instead of `update_user_metadata_and_profile` hook is more confusing.